Adding Nix tooling for build & development

efim opened the pull request.

efim committed:
Add Nix tools for build & development

Nix is a fully functional package management
where system can be configured in a fully declarative way.

It is already easy to install Searx via Nix, by adding a couple of lines
into the server configuration, this commit allows for installation of
the plugin.

return42 commented:
Thanks for your contribution / FYI

- I do prefer [SearXNG](https://github.com/searxng/searxng/tree/master) over the origin searX.
- In the SearXNG project we active maintain the scripts in utils/ (compare https://github.com/searxng/searxng/tree/master/utils vs https://github.com/searx/searx/tree/master/utils)
- The commits from your PR are tagged "WIP" .. if you think PR is ready for review please git squash 
- I'm not familiar with Nix .. if you can give some hints about how I can test this PR would be great
-

efim commented:
Hello!

I didn't mean to open this PR yet (this was an unplanned mistake on my part), I'm still in finalizing discussion with mentor in "Summer of Nix" programm

After that discussion I'll add more notes & documentation on any points that might be relevant

efim commented:
Nix is a purely functional package manager & NixOS is a GNU\Linux distribution built with that package manager
NixPkgs provides packages for any system running the Nix package
manager (available on Linux & MacOS).

It allows to specify full system (server, container or personal machine) declaratively

Searx is already packaged in nixpkgs, and can be installed onto NixOS server by adding couple lines into configuration
this work is supposed to allow installation of the external plugins into searx

Here is the example I've created of a configuration that would set up Searx instance
and include the "external plugins" - by using the `flake.nix` which I've written for the code in this repository
https://github.com/efim/nix-searx-container-example

And I will definitely squash the commits before submitting contribution to your repository

gytis-ivaskevicius commented on flake.nix:
```diff
@@ -0,0 +1,109 @@
+    owner = "NixOS";
+    repo = "nixpkgs";
+    ref = "21.05";
+  };
```
How about:
```suggestion
  inputs.nixpkgs.url = "github:NixOS/nixpkgs/21.05";
```
Or maybe even:
```suggestion
  inputs.nixpkgs.url = "github:NixOS/nixpkgs/nixos-unstable-small";
```

efim replied:
With "nixos-unstable-small" something doesn't work during the plugin build, but I'm not too inclined to investigate, 
21.05 seems recent enough

gytis-ivaskevicius commented on flake.nix:
```diff
@@ -0,0 +1,109 @@
+              prev.searx
+            ];
+            src = ./.;
+            doCheck = false;
```
What are the issues with tests? It would be nice to cover it at least by some sort of tests

efim replied:
I've enabled tests, but couldn't figure out how to check whether anything useful is being executed during bulid process =C

gytis-ivaskevicius commented on flake.nix:
```diff
@@ -0,0 +1,109 @@
+          searx = prev.searx.overrideAttrs (oldAttrs: {
+              propagatedBuildInputs = oldAttrs.propagatedBuildInputs ++ [
+                tgwf-green-results-searx-plugin
+              ];
```
Are you sure that it is working as expected?

efim replied:
tldr; works with `propagatedBuildInputs`, doesn't work with just `buildInputs`

I've tested with container running Searx instance
If I change that to `buildInputs` Searx doesn't start because it can't see the plugin installed

I'm not totally sure about the inner workings, but it seems that the plugin has to be installed in pip and looks like `propagatedBuildInputs` is what's needed (as I understand it lists packages needed at runtime)

gytis-ivaskevicius commented on flake.nix:
```diff
@@ -0,0 +1,109 @@
+          {
+            inherit (nixpkgsFor.${system}) tgwf-green-results-searx-plugin searx;
+          }
+        );
```
Consider reducing amount of whitespace like so:
```suggestion
        packages = forAllSystems (system: {
            inherit (nixpkgsFor.${system}) tgwf-green-results-searx-plugin searx;
        });
```

efim replied:
I would agree with that formatting, `nixpkgs-fmt` disagrees
I propose to use auto-formatters output, that would result in more stable diffs

gytis-ivaskevicius commented on flake.nix:
```diff
@@ -0,0 +1,109 @@
+
+            plugins:
+              - only_show_green_results
+            '';
```
Consider such indentation:
```suggestion
          settingsFile = nixpkgsFor.${system}.writeText "settings.yml" ''
            use_default_settings: True
            server:
                secret_key : "dev-server-secret"

            plugins:
              - only_show_green_results
          '';
```

Also, since this is a devshell maybe it would make more sense to have this file committed within repo and used from there?

efim replied:
I will in the future look at the possible pros and cons to separating this configuration part into separate file

One possible significant bonus from keeping it in the `flake.nix` is more ease with refacotirng,
since the configuration text references 
```
                plugins:
                  - ${pluginName}
```
in the refactored version, so that there aren't many hardcoded "only_show_green_results"

gytis-ivaskevicius commented on .gitignore:
```diff
@@ -1,4 +1,6 @@
\ No newline at end of file
+tgwf_searx_plugins.egg-info
+/outputs/
+/result
```
```suggestion
result
```

gytis-ivaskevicius reviewed (approved):
Good stuff! :+1:

efim committed:
Add readme

efim commented:
Hello @return42  !
I've cleaned up changes, and added notes into Readme file

My hope is that now it for anyone who runs Searx via Nix it will be very easy to install the TGWF plugin.

I would want the notes & lined sample configuration to be absolutely clear for anyone who knows Nix and quite clear for those who never used it before.
Please let me know if there's anything you'd want to know or would like added

I will look into `SearXNG`, hopefully it is similar enough to `SearX` so that I'd be able to reuse Nix code for it
and people would be able to install & configure it with Nix as well (and for same "external plugin" system to work for plugins)

return42 commented:
> I will look into SearXNG, hopefully it is similar enough to SearX so that I'd be able to reuse Nix code for it

As I said before ..

> In the SearXNG project we active maintain the scripts in utils/ (compare https://github.com/searxng/searxng/tree/master/utils vs https://github.com/searx/searx/tree/master/utils)

That is why I recommend to use [SearXNG](https://github.com/searxng/searxng/) in _Nix_.

efim commented:
> In the SearXNG project we active maintain the scripts in utils/ (compare https://github.com/searxng/searxng/tree/master/utils vs https://github.com/searx/searx/tree/master/utils)

Could you clarify that a bit, I might not fully understand the thread
Do you mean that in the `SearXNG` project the team invests noticeable effort in updating the utils and that is a great argument for having `SearXNG` packaged for Nix

And it doesn't mean that changes to utils change how two projects interact with the external plugins?
i.e if I understand correctly the external plugins for now have identical installation process for both `SearXNG` and `Searx`?

And by the previous comment I mean that I will look into creating a package for the `SearXNG` for Nix
Hovewer even if I'm able to do that - it would be a PR into the `SearXNG` with the `flake.nix` that would allow installation & use of dev env for `SearXNG` and doesn't depend on this PR

return42 commented:
Sorry, I mixed plugin and SearX / SearXNG ..

> Could you clarify that a bit, I might not fully understand the thread
> Do you mean that in the `SearXNG` project the team invests noticeable effort in updating the utils and that is a great argument for having `SearXNG` packaged for Nix

Depends on, you are free to use whatever fork you prefer.   I do not know how package installer in Nix are implemented, if the installer uses the [Installation scripts](https://searxng.github.io/searxng/admin/installation.html#installation-scripts) from `/utils` I just want to point out, that these scripts do have some issues and had been rewritten slightly in the SearXNG fork.

SearXNG is a fork of Searx, both branches profit from each other and from the same community.  But over the time both branches differs more and more and I can't speak for compatibility issues in the fututre (see [README.rst](https://github.com/searxng/searxng/blob/master/README.rst))

If I'm not wrong, ATM the plugins API should be compatible between SearX & SearXNG.

> And by the previous comment I mean that I will look into creating a package for the `SearXNG` for Nix

This would be a great addition, I know that @dalf and me are looking forward for Nix support. Do you like to add a PR at SerXNG.

BTW where can I find what reposetory is cloned to build the instance? I looked at https://github.com/efim/nix-searx-container-example/blob/master/flake.nix may be the wrong file?

efim commented:
I want to start by stating that I wish success to both projects!
I found this repo because I was working specifically on the TGWF plugin, and got acquanted with SearXNG from your messages.

I might not fully understand the question about building an instance
I'll write about both possibile interpretations that I saw:

### If you want to check out full container that runs Searx + plugin:
The link to example server configuration that sets up Searx & the plugin is correct:
- for stable version of nix (without flakes) this config: https://github.com/efim/nix-searx-container-example/blob/master/non-flake-config.nix

- for nix with new helpful experimental feature of flakes: https://github.com/efim/nix-searx-container-example/blob/master/flake.nix

These configs can be used on NixOS to start a container with Searx installed and the plugin applied.
They are examples of what admins would have to do to have Searx with the plugin running on their servers.
I tried to write up guide on how to set it up & troubleshoot on NixOS in the readme for that repo

To run Searx as systemd service on a server that uses NixOS configuration looks pretty much the same, part about Searx can be just copied.

I don't know much about running services managed by nix on non-NixOS, packages can be build and run on any OS, but modules that describe `systemd` services and create configuration files might not be that easy to use without NixOS.

###  If you want only to build Searx with the plugin and check how it works
Then any OS is good and you would only need nix package manager & the repository of this plugin.

With nix installed running
`nix build` will build the Searx with the plugin added

result of the build will be awailable in the ./results/bin/searx-run
(only configuration with non-default values would have to be provided for it to start)

alternatively command
`nix-shell` (without flakes) or `nix develop` (with flakes)
will not only build the Searx + plugin, but also enter new shell that has variable SEARX_SETTINGS_PATH set to simple config file & "searx-run" on the path

So locally only entering the `nix-shell` is needed to launch the Searx with plugin to check how it works

return42 commented:
@efim thanks a lot for your explanations .. very helpful for me .. I think have to learn some basic about Nix and NixOS :)
